### PR TITLE
DAOS-2644 control: prompt to respawn under new user after chown files

### PR DESCRIPTION
**Pull request by tanabarr**
Dropping privileges of current running process doesn't work because SPDK
resources can't be released and reinitialised within process lifetime.

In order to work around this issue, when running as root, after format;
change permissions of required files and alert the user to respawn as
user specified in server config file, then exit. Also minor related fixups.

  - change file ownership and log message to respawn under new user
  - only register minimal gRPC services if going to need to respawn
  - recognise missing config log file location parameter
  - don't show confirmation text if --force applied to dmg storage format
  - separate createSocketDir from drpcSetup

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/server/config.go [hunk 1]
```diff
@@ -108,7 +108,8 @@ func loadConfigOpts(cliOpts *cliOptions, host string) (
 	log.Debugf("DAOS config read from %s", config.Path)
 
 	// get unique identifier to activate SPDK multiprocess mode
-	config.NvmeShmID = hash(host + strconv.Itoa(os.Getpid()))
+	// TODO: re-enable
+	//config.NvmeShmID = hash(host + strconv.Itoa(os.Getpid()))
```
Comment (liw) [inline src/control/server/config.go]: Is this change related to the purpose of this patch?

File: src/control/server/drpc.go [hunk 2]
```diff
@@ -46,17 +46,26 @@ func getDrpcClientConnection(sockDir string) *drpc.ClientConnection {
 // drpcSetup creates socket directory, specifies socket path and then
 // starts drpc server.
 func drpcSetup(sockDir string, iosrv *iosrv) error {
+	var msg string
+
```
Comment (liw) [inline src/control/server/drpc.go]: msg could be restricted to the scope of "{" on new line 53.

File: src/control/server/main.go [hunk 3]
```diff
@@ -123,32 +122,39 @@ func serverMain() error {
 	go grpcServer.Serve(lis)
 	defer grpcServer.GracefulStop()
 
+	if err := mgmtControlServer.nvme.spdk.prep(
+		1024, config.UserName, ""); err != nil {
```
Comment (liw) [inline src/control/server/main.go]: To be honest, I always find this line-breaking style weird. Is it idiomatic Go?

File: src/control/server/privileges.go [hunk 4]
```diff
@@ -39,14 +43,20 @@ func getUID(ext External, userName string) (*user.User, int64, error) {
...
 
 	return usr, uid, nil
 }
 
-func getGID(ext External, usr *user.User, groupName string) (int64, error) {
+// getGID returns group name and id, either of the supplied group if user
+// belonga to group or of usr otherwise.
+func getGID(
+	ext External, usr *user.User, groupName string) (
+	grpName string, gid int64, err error) {
```
Comment (liw) [inline src/control/server/privileges.go]: Same comments on the line-breaking style...

File: src/control/server/privileges.go [hunk 5]
```diff
@@ -66,38 +77,53 @@ func getGID(ext External, usr *user.User, groupName string) (int64, error) {
...
+	// add last because we don't want to fail to write before chown proc
+	paths = append(paths, config.ControlLogFile)
+
+	for _, path := range paths {
+		if path == "" {
+			continue
+		}
+
+		err := config.ext.runCommand(
+			fmt.Sprintf("chown -R %s:%s %s", user, group, path))
```
Comment (liw) [inline src/control/server/privileges.go]: I saw recommendations for filepath.Walk. That would still be too complex?

File: src/control/server/privileges.go [hunk 6]
```diff
@@ -66,38 +77,53 @@ func getGID(ext External, usr *user.User, groupName string) (int64, error) {
...
+	paths := []string{
+		config.SocketDir,
+	}
+
+	for _, srv := range config.Servers {
+		paths = append(paths, srv.ScmMount, srv.LogFile)
+	}
+
+	// add last because we don't want to fail to write before chown proc
+	paths = append(paths, config.ControlLogFile)
```
Comment (liw) [inline src/control/server/privileges.go]: Would it be easier to maintain if we write chown in the code near the code that create the files?

File: src/control/server/drpc.go [hunk 7]
```diff
@@ -46,17 +46,26 @@ func getDrpcClientConnection(sockDir string) *drpc.ClientConnection {
 // drpcSetup creates socket directory, specifies socket path and then
 // starts drpc server.
 func drpcSetup(sockDir string, iosrv *iosrv) error {
```
Comment (mjmac) [inline src/control/server/drpc.go]: It would be simpler to break out the management of the sockDir into its own function which is solely concerned with the sockDir. Following the pattern of returning as early as possible also tends to shorten and simplify code paths.

```go
func createSocketDir(sockDir string) error {
	// FIXME: shouldn't there be some checks on the socket dir even if it exists?
	// e.g. permissions, that it's actually a directory and not a file, etc?
	_, err := os.Stat(sockDir)
	if err == nil {
		return nil
	}

	switch {
	case os.IsPermission(err):
		return errors.Wrapf(err, "user does not have permission to access %s", sockDir)
	case os.IsNotExist(err):
		if err := os.MkdirAll(sockDir, 0755); err != nil {
			return errors.Wrapf(err, "unable to create socket directory %s", sockDir)
		}
	}

	return errors.Wrapf(err, "unknown error while creating sockDir %s", sockDir)
}

// drpcSetup creates socket directory, specifies socket path and then
// starts drpc server.
func drpcSetup(sockDir string, iosrv *iosrv) error {
	if err := createSocketDir(sockDir); err != nil {
		// probably no value added by wrapping here
		return err
	}
...
```

Comment (tanabarr) [reply src/control/server/drpc.go]: sounds good but maybe should be in its own patch? if you agree could you create a ticket and just paste this comment please

File: src/control/server/privileges.go (hunk 5 above)
Comment (tanabarr) [reply src/control/server/privileges.go]: that in order to use native Go Api or for some other reason?

**Comment by tanabarr** [comment]
re-factoring because of some fundamental issues regarding the approach and SPDK process ownership

File: src/control/server/privileges.go (hunk 5 above)
Comment (liw) [reply src/control/server/privileges.go]: To avoid executing the chown program. This is an open question. If using filepath.Walk would still be too complex, then let's forget about it.

File: src/control/server/config.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/config.go]: related but experimental only, removed

File: src/control/server/main.go (hunk 3 above)
Comment (tanabarr) [reply src/control/server/main.go]: removed

File: src/control/server/privileges.go (hunk 4 above)
Comment (tanabarr) [reply src/control/server/privileges.go]: I've seen line breaking performed like this in various style guides, I tend to prefer line breaks after opening parenthesis. is there another way you would prefer to do this?

File: src/control/server/privileges.go (hunk 5 above)
Comment (tanabarr) [reply src/control/server/privileges.go]: no I think that would be fine

File: src/control/server/drpc.go (hunk 2 above)
Comment (tanabarr) [reply src/control/server/drpc.go]: done

File: src/control/server/privileges.go (hunk 6 above)
Comment (tanabarr) [reply src/control/server/privileges.go]: could you clarify, do you mean adding the files to a list of "files to chown" at the code where we create the files?

File: src/control/server/drpc.go (hunk 7 above)
Comment (tanabarr) [reply src/control/server/drpc.go]: nvm, made the changes

**Comment by tanabarr** [comment]
verified as working on boro, need to fix up unit tests

File: src/control/server/privileges.go (hunk 6 above)
Comment (liw) [reply src/control/server/privileges.go]: If possible, the best way might be

    create file 0, 1, 2
    chown file 0, 1, 2
    ...
    create file 3, 4
    chown file 3, 4
    ...

Maintaining a list "files to chown" might be second best.

**Comment by tanabarr** [comment]
gatekeeper please use PR title and description for commit message on merge

File: src/control/server/ownership.go [hunk 8]
```diff
@@ -135,3 +131,32 @@ func changeFileOwnership(config *configuration) error {
 
 	return nil
 }
+
+// respawnProc builds and executes command chain to respawn process under new ownership.
+func respawnProc(config *configuration) error {
+	var buf bytes.Buffer // build command string
+
+	// Wait for this proc to exit and make NVMe storage accessible to new user.
+	fmt.Fprintf(
```
Comment (mjmac) [inline src/control/server/ownership.go]: style: this newline is unnecessary

File: src/control/server/ownership.go [hunk 9]
```diff
@@ -135,3 +131,32 @@ func changeFileOwnership(config *configuration) error {
...
+		&buf, `sleep 1 && %s storage prep-nvme -u %s &> %s`,
+		os.Args[0], config.UserName, config.ControlLogFile)
+
+	// Run daos_server from within a subshell of target user with the same args.
+	fmt.Fprintf(&buf, ` && su %s -c "`, config.UserName)
+
+	for _, arg := range os.Args {
```
Comment (mjmac) [inline src/control/server/ownership.go]: you could also do this: 
```go
fmt.Fprintf(&buf, "%s", strings.Join(os.Args))
```
or, probably better, if you use strings.Builder you could simplify a lot of stuff here and avoid using fmt.Fprintf:

```go
buf.WriteString(strings.Join(os.Args, " "))

File: src/control/server/ownership.go [hunk 10]
```diff
@@ -135,3 +131,32 @@ func changeFileOwnership(config *configuration) error {
 
 	return nil
 }
+
+// respawnProc builds and executes command chain to respawn process under new ownership.
+func respawnProc(config *configuration) error {
```
Comment (mjmac) [inline src/control/server/ownership.go]: Why are these two very different operations stuck together?

From a clarity, testing, and maintenance perspective, it seems better to have a function which will do the NVMe prep, then a separate function to fork/exec with lowered privileges.

File: src/control/server/server.go [hunk 11]
```diff
@@ -116,11 +114,13 @@ func Main() error {
 
 	mgmtpb.RegisterMgmtCtlServer(grpcServer, mgmtCtlSvc)
 
-	// Only provide IO/Agent communication if running as non-root user or if no
-	// non-root user has been specified in config (not recommended, will log error).
-	// If root, only provide gRPC mgmt control service for hardware provisioning.
-	if syscall.Getuid() != 0 || config.UserName == "" {
-		mgmtpb.RegisterMgmtSvcServer(grpcServer, newMgmtSvc(&config))
+	// If running as root and user name specified in config file, respawn proc.
+	respawn := syscall.Getuid() == 0 && config.UserName != ""
```
Comment (mjmac) [inline src/control/server/server.go]: suggestion: shouldRespawn is extremely clear and retains clarity regardless of how far away the variable is used from its declaration.

File: src/control/server/config.go [hunk 12]
```diff
@@ -93,37 +93,35 @@ func (c *configuration) setPath(path string) error {
...
+	config := newConfiguration()
 
 	if err := config.setPath(cliOpts.ConfigPath); err != nil {
-		return config, errors.WithMessage(err, "set path")
+		return &config, errors.WithMessage(err, "set path")
 	}
 
 	if err := config.loadConfig(); err != nil {
-		return config, errors.Wrap(err, "read config file")
+		return &config, errors.Wrap(err, "read config file")
```
Comment (mjmac) [inline src/control/server/config.go]: Shouldn't these be returning nil, errors.Wrap(err, ...)? If failing to perform these steps are errors, then returning a config after these failures could result in undefined behavior at some point if the caller doesn't handle errors as you expect here.

File: src/control/server/ownership.go (hunk 10 above)
Comment (tanabarr) [reply src/control/server/ownership.go]: we need to do both after this process dies, therefore they need to be change together and executed independently after this process has been able to exit

Comment (mjmac) [reply src/control/server/ownership.go]: Please elaborate on this. Why does storage prep have to happen after the parent daos_server process dies?

Comment (tanabarr) [reply src/control/server/ownership.go]: because the handles that the running process has open on SPDK libraries prevent storage-prep from doing what needs to and those references can't be completely cleaned up during the process lifetime.

Comment (tanabarr) [reply src/control/server/ownership.go]: more details in https://jira.hpdd.intel.com/browse/DAOS-2644

File: src/control/server/ownership.go (hunk 9 above)
Comment (tanabarr) [reply src/control/server/ownership.go]: in the other cases, is `buf.WriteString(fmt.Sprintf(` && su %s -c "`, config.UserName))` simpler than `fmt.Fprintf(&buf, ` && su %s -c "`, config.UserName)` or is there another way of doing it?

File: src/control/server/drpc.go [hunk 13]
```diff
@@ -43,21 +46,32 @@ func getDrpcClientConnection(sockDir string) *drpc.ClientConnection {
 	return drpc.NewClientConnection(clientSock)
 }
 
+func createSocketDir(sockDir string) error {
```
Comment (dpquigl) [inline src/control/server/drpc.go]: Im ok with this for now but I have been debating if we should even be doing this. Since the socket directories are by default under a system controlled tmpfs maybe we just rely on the unit files or the rpm to make the right directories and we fail if its not there.

Comment (tanabarr) [reply src/control/server/drpc.go]: yes I remember that discussion but was deferring making the decision as I don't think it's part of this patch
